### PR TITLE
bpf: Adds support for drop IPv4 fragmented packet

### LINKED ISSUE #9894: Add metric for dropped packets due to IP fragmentation
Cilium datapath does not support IP fragmentation, and fragmented packets are usually dropped.

To make users aware of such cases, we need to export a metric for dropped packets. This requires adding `return DROP_FRAG_NOSUPPORT;` to the datapath and adding an adequate reason to the `cilium_drop_{count,bytes}_total` metrics.

### PR DESCRIPTION
**Pull request by navarrothiago**
If IPv4 fragmentation is disabled and a IPv4 fragmented packet is received,
then the packet is dropped (`DROP_FRAG_NOSUPPORT`).

Due to this return error, the metrics `cilium_drop_{count,bytes}_total`
are updated.

Besides, fixes IPv4 fragment tracking option in [documentation](https://docs.cilium.io/en/v1.9/concepts/networking/fragmentation/#concepts-fragmentation).

Fixes: #9894

Signed-off-by: Thiago Navarro navarro@accuknox.com

### CODE REVIEW

**Review by qmonnet — APPROVED** [review]
Looks good to me, thanks a lot! Nice catch for the typo in the docs.

**Comment by qmonnet** [comment]
test-me-please

**Comment by qmonnet** [comment]
test-1.16-netnext

**Comment by qmonnet** [comment]
k8s-1.16-kernel-netnext hit https://github.com/cilium/cilium/issues/15737, which is not related.

**Comment by navarrothiago** [comment]
Should we retrigger the job that failed? The #15737 was already merged.

**Comment by qmonnet** [comment]
No it's not necessary, we're just waiting for the review from @jrfastab.

**Review by pchaigno — APPROVED** [review]

**Comment by pchaigno** [comment]
@qmonnet Do we have tests running without fragment tracking support? I'm wondering how well this is covered with regards to complexity issues in our datapath.

**Comment by qmonnet** [comment]
@pchaigno No I don't think we have. Fragment tracking is enabled by default and I can't remember of or find any test which would disable it.

**Comment by pchaigno** [comment]
It might be good to run the three K8sVerifier tests locally without `ENABLE_IPV4_FRAGMENTS` in `bpf/Makefile` to check we're not adding more complexity issues with that pull request.

**Comment by navarrothiago** [comment]
@pchaigno, I have removed the `ENABLE_IPV4_FRAGMENTS` in `bpf/Makefile` and executed the test with this command
```
NETNEXT=1 ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=false --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
```
Find attached the diff and the output. 
[logfile_2021-04-23-133822.txt](https://github.com/cilium/cilium/files/6366892/logfile_2021-04-23-133822.txt)
[removes-enable-ipv4-fragments.txt](https://github.com/cilium/cilium/files/6366893/removes-enable-ipv4-fragments.txt)

Please, let me know if it was what you have addressed.

**Comment by pchaigno** [comment]
@navarrothiago Oh nice, I didn't expect you to do it yourself. I should have clarified :-/

We will also need to test the following two other kernels with the same diff:
```
ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=false --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
KERNEL=419 ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=false --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
```
If you don't have time, I can run those next week.

**Comment by navarrothiago** [comment]
@pchaigno, I have done what you suggested and bingo! It failed - `BPF program is too large. Processed 1000001 insn`. Is it supposed to happen? If so,  I could check I way how to reduce the program or increase the limit. Any suggestion? 

Command
```
ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=false --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
```
:x: Output - [logfile_2021-04-23-140202.txt](https://github.com/cilium/cilium/files/6367389/logfile_2021-04-23-140202.txt)

---
Command
```
KERNEL=419 ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=false --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
```
:x: Output - [logfile_2021-04-23-152839.txt](https://github.com/cilium/cilium/files/6367418/logfile_2021-04-23-152839.txt)

**Comment by pchaigno** [comment]
That's weird. I just realized the first shouldn't have changed anything since fragment tracking is disabled on 4.9 (the default kernel when using ginkgo for e2e tests). Maybe you didn't destroy the VMs before running these two tests? (They run on different kernels so you need to destroy before relaunching.)

**Comment by navarrothiago** [comment]
@pchaigno Thank you! You are right. I didn't do that Good to know.

**Comment by pchaigno** [comment]
Fragment tracking is not supported 4.9 because of the lack of support for some BPF maps. So it can't be enabled there :-)

**Comment by navarrothiago** [comment]
@pchaigno, here is the logs as suggested with the same diff.

Command
```
ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=true --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
```
:heavy_check_mark: : Output - [logfile_2021-04-25-200200.txt](https://github.com/cilium/cilium/files/6373409/logfile_2021-04-25-200200.txt)


---
Command
```
KERNEL=419 ginkgo --focus="K8s.*Verifier" -v -- --cilium.provision=true --cilium.skipLogs=true -cilium.showCommands=true | tee logs/logfile_$(date '+%Y-%m-%d-%H%M%S').txt
```
:heavy_check_mark: Output - [logfile_2021-04-25-194123.txt](https://github.com/cilium/cilium/files/6373395/logfile_2021-04-25-194123.txt)

**Comment by pchaigno** [comment]
@navarrothiago Thanks for running those! :pray: 

This is ready to merge, flagging as such.

**Pull request merged**
